app/clients/TrackDownloadsClient.ts

```diff
@@ -0,0 +1,254 @@
+/** Represents a track which hasn't been downloaded yet.
+ */
+type DownloadStateNotDownloaded = { status: "NOT_DOWNLOADED" }
+
+export type DownloadState =
+  | DownloadStateUnknown
+  | DownloadStateDownloaded
+  | DownloadStateDownloading
+  | DownloadStatePaused
+  | DownloadStateFailed
+  | DownloadStateNotDownloaded
+
+class TrackDownloadsClient {
+  private cachedDownloadState: Map<string, DownloadState>
+  private listeners: Map<string, (() => void)[]>
+
+  constructor() {
+    this.cachedDownloadState = new Map()
+    this.listeners = new Map()
+  }
+
+  findDownloadInLocalFileSystem(track: Track) {
+    const entry = this.fetchDownloadState(track.trackId)
+
+    if (entry.status !== "UNKNOWN") {
+      return
+    }
+
+    FileSystem.getInfoAsync(getTrackFileUri(track), { md5: true }).then(
+      (fileInfo) => {
+        if (fileInfo.exists) {
+          this.cachedDownloadState.set(track.trackId, {
+            status: "DOWNLOADED",
+            uri: fileInfo.uri,
+            md5: fileInfo.md5!,
+          })
+          this.notify(track.trackId)
+        } else {
+          this.cachedDownloadState.set(track.trackId, {
+            status: "NOT_DOWNLOADED",
+          })
+          this.notify(track.trackId)
+        }
+      },
+      (error) => {
+        console.error(`Failed to determine download status of track ${track.trackId}.`, error)
+      },
+    )
+  }
+
+  fetchDownloadState(trackId: string): DownloadState {
+    const downloadState = this.cachedDownloadState.get(trackId)
+
+    if (downloadState === undefined) {
+      return { status: "UNKNOWN" }
+    } else {
+      return downloadState
+    }
+  }
+
+  startOrResumeDownload(track: Track): void {
+    const currentState = this.fetchDownloadState(track.trackId)
+
+    // TODO: Handle download failed case.
+    if (currentState.status === "NOT_DOWNLOADED" || currentState.status === "FAILED_DOWNLOADING") {
+      const callback = ({ totalBytesWritten, totalBytesExpectedToWrite }) => {
+        const state = this.fetchDownloadState(track.trackId)
+        if (state.status === "DOWNLOADING" || state.status === "DOWNLOAD_PAUSED") {
+          state.progress = totalBytesWritten / totalBytesExpectedToWrite
+        }
+        this.notify(track.trackId)
+      }
+
+      const downloadResumable = FileSystem.createDownloadResumable(
+        track.webUri,
+        getTemporaryFileUri(track),
+        // NOTE: Make sure to account for FileSystemDownloadResult.md5 not
+        // existing if you change this to `false` or remove it.
+        { md5: true },
+        callback,
+      )
+      downloadResumable.downloadAsync().then(
+        (r) => {
+          this.handleDownloaded(track, r)
+        },
+        (e) => {
+          this.handleDownloadFailed(track.trackId, e)
+        },
+      )
+      this.cachedDownloadState.set(track.trackId, {
+        status: "DOWNLOADING",
+        progress: 0,
+        downloadResumable,
+      })
```

Comment:
If I understand JS's execution correctly a "too fast promise resolve" can't happen since JS is executed in a single thread and the runtime has to wait until the current "task" gives control back in order to make progress. This would happen at the end of this function (https://en.wikipedia.org/wiki/Cooperative_multitasking#Usage). So, `this.handleDownloaded` is never be called before `this.cacheDownloadState.set(track.trackId), ...)`.

The control flow is indeed not easy to follow, though. I'll try to use `await`. This would change the return type from `void` to `Promise<?>`, so users of the API could "subscribe" to finished downloads (`subscribe` would already give you that). That's something I wanted to avoid since I don't always create a promise (e.g. if the download is already running, would need to return a dummy promise in this case) and the function starts two promises both with potentially conflicting resolve types.

Reply:
Ah, thanks for the explanation of the control flow. Could a change of the return type be avoided by having some anonymous `async` function that uses `await` but is called without `await`. Like what you do when you want to use `await` in `useEffect`? I don't know though if that really makes it clearer what is happening. I'm happy with the current implementation and your control flow explanation.

Reply:
Yes, that is what I had in mind. I don't find it more clear and readable though. Maybe the first version with setting the status before invoking `downloadAsync` is better. Do whatever feels good to you; it seems to me to be more about aesthetics than code quality. Maybe some consistency would be good though.